File: saizad-mvvm/src/main/java/com/saizad/mvvm/SaizadEasyRetrofit.java
package com.saizad.mvvm;

import android.app.Application;

import androidx.annotation.NonNull;

import com.google.gson.Gson;

import retrofit2.Converter;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;
import sa.zad.easyretrofit.EasyRetrofit;
import sa.zad.easyretrofit.EasyRetrofitClient;

public class SaizadEasyRetrofit extends EasyRetrofit {

    private final CurrentUser currentUser;
    private final Gson gson;
    private final String domainUrl;

    public SaizadEasyRetrofit(Application application, CurrentUser currentUser, Gson gson, String domainUrl) {
        super(application);
        this.currentUser = currentUser;
        this.gson = gson;
        this.domainUrl = domainUrl;
    }

    @NonNull
    @Override
    public Retrofit.Builder retrofitBuilderReady(@NonNull Retrofit.Builder retrofitBuilder) {
        return retrofitBuilder
                .baseUrl(domainUrl);
    }

    @NonNull
    @Override
    protected Converter.Factory addConverterFactory() {
        return GsonConverterFactory.create(gson);
    }

    @NonNull
    @Override
    protected EasyRetrofitClient easyRetrofitClient() {
        return new SaizadEasyRetrofitClient(provideApplication(), currentUser);
    }

}

File: saizad-mvvm/src/main/java/com/saizad/mvvm/SaizadEasyRetrofit.kt
package com.saizad.mvvm

import android.app.Application
import com.google.gson.Gson
import retrofit2.Converter
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import sa.zad.easyretrofit.EasyRetrofit
import sa.zad.easyretrofit.EasyRetrofitClient

class SaizadEasyRetrofit(
    protected val application: Application,
    protected val currentUser: CurrentUserType<*>,
    protected val gson: Gson,
    protected val domainUrl: String,
    val isDebugMode: Boolean = false
) : EasyRetrofit(application) {

    override fun retrofitBuilderReady(retrofitBuilder: Retrofit.Builder): Retrofit.Builder {
        return retrofitBuilder
            .baseUrl(domainUrl)
    }

    override fun addConverterFactory(): Converter.Factory {
        return GsonConverterFactory.create(gson)
    }

    override fun easyRetrofitClient(): EasyRetrofitClient {
        return SaizadEasyRetrofitClient(provideApplication(), currentUser, isDebugMode)
    }

}